File: querydsl-core/src/test/java/com/mysema/query/codegen/EntitySerializerTest.java
/*
 * Copyright (c) 2010 Mysema Ltd.
 * All rights reserved.
 *
 */
package com.mysema.query.codegen;

import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.io.StringWriter;
import java.sql.Time;
import java.util.Collections;
import java.util.Date;

import org.junit.Test;

import com.mysema.codegen.JavaWriter;
import com.mysema.codegen.model.ClassType;
import com.mysema.codegen.model.SimpleType;
import com.mysema.codegen.model.TypeCategory;
import com.mysema.query.annotations.PropertyType;

public class EntitySerializerTest {

    EntitySerializer serializer = new EntitySerializer(new TypeMappings(), Collections.<String>emptySet());
    StringWriter writer = new StringWriter();
    
    @Test
    public void No_Package() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType("Q",type);        
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("public class QEntity extends EntityPathBase<Entity> {"));
    }
    
    @Test
    public void Correct_Superclass() throws IOException {
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "java.util.Locale", "java.util", "Locale",false,false);
        EntityType entityType = new EntityType("Q",type);        
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        System.out.println(writer);
        assertTrue(writer.toString().contains("public class QLocale extends EntityPathBase<java.util.Locale> {"));
    }
    
    @Test
    public void Primitive_Array() throws IOException{
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType("Q",type);
        entityType.addProperty(new Property(entityType, "bytes", new ClassType(byte[].class)));
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        assertTrue(writer.toString().contains("public final SimplePath<byte[]> bytes"));
    }
    
    @Test
    public void Include() throws IOException{
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType("Q",type);
        entityType.addProperty(new Property(entityType, "b", new ClassType(TypeCategory.BOOLEAN, Boolean.class)));
        entityType.addProperty(new Property(entityType, "c", new ClassType(TypeCategory.COMPARABLE, String.class)));
        entityType.addProperty(new Property(entityType, "cu", new ClassType(TypeCategory.CUSTOM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "d", new ClassType(TypeCategory.DATE, Date.class)));
        entityType.addProperty(new Property(entityType, "e", new ClassType(TypeCategory.ENUM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "dt", new ClassType(TypeCategory.DATETIME, Date.class)));
        entityType.addProperty(new Property(entityType, "i", new ClassType(TypeCategory.NUMERIC, Integer.class)));
        entityType.addProperty(new Property(entityType, "s", new ClassType(TypeCategory.STRING, String.class)));
        entityType.addProperty(new Property(entityType, "t", new ClassType(TypeCategory.TIME, Time.class)));
        
        EntityType subType = new EntityType("Q",new SimpleType(TypeCategory.ENTITY, "Entity2", "", "Entity2",false,false));
        subType.include(new Supertype(type,entityType));
        
        serializer.serialize(subType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        // TODO : assertions
    }
    
    @Test
    public void Properties() throws IOException{
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType("Q",type);
        entityType.addProperty(new Property(entityType, "b", new ClassType(TypeCategory.BOOLEAN, Boolean.class)));
        entityType.addProperty(new Property(entityType, "c", new ClassType(TypeCategory.COMPARABLE, String.class)));
        entityType.addProperty(new Property(entityType, "cu", new ClassType(TypeCategory.CUSTOM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "d", new ClassType(TypeCategory.DATE, Date.class)));
        entityType.addProperty(new Property(entityType, "e", new ClassType(TypeCategory.ENUM, PropertyType.class)));
        entityType.addProperty(new Property(entityType, "dt", new ClassType(TypeCategory.DATETIME, Date.class)));
        entityType.addProperty(new Property(entityType, "i", new ClassType(TypeCategory.NUMERIC, Integer.class)));
        entityType.addProperty(new Property(entityType, "s", new ClassType(TypeCategory.STRING, String.class)));
        entityType.addProperty(new Property(entityType, "t", new ClassType(TypeCategory.TIME, Time.class)));
        
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        // TODO : assertions
    }
    
    @Test
    public void SuperType() throws IOException{
        EntityType superType = new EntityType("Q",new SimpleType(TypeCategory.ENTITY, "Entity2", "", "Entity2",false,false));
        SimpleType type = new SimpleType(TypeCategory.ENTITY, "Entity", "", "Entity",false,false);
        EntityType entityType = new EntityType("Q",type, Collections.singleton(new Supertype(superType, superType)));
        
        serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new JavaWriter(writer));
        // TODO : assertions
    }

}

File: querydsl-scala/src/test/scala/com/querydsl/scala/EntitySerializerTest.scala
package com.querydsl.scala

import com.querydsl.core.types._

import org.junit.{ Test, Before, After, Assert }
import org.junit.Assert._

import com.mysema.codegen._
import com.mysema.codegen.model._

import com.querydsl.codegen._
import com.querydsl.sql._

import java.io.StringWriter

class ScalaEntitySerializerTest extends CompileTestUtils {

  var entityType: EntityType = null

  val writer = new StringWriter()

  @Before
  def setUp() {
    val typeModel = new ClassType(TypeCategory.ENTITY, classOf[Person])
    entityType = new EntityType(typeModel)
    for ( (name, t) <- List(
        ("scalaInt", Types.INT),
        ("javaInt", Types.INTEGER),
        ("javaDouble", Types.DOUBLE),
        ("scalaDouble", Types.DOUBLE_P),
        ("firstName", Types.STRING),
        ("lastName", Types.STRING),
        ("scalaList", new SimpleType(Types.LIST, Types.STRING)),
        ("scalaMap", new SimpleType(Types.MAP, Types.STRING, Types.STRING)),
        ("javaCollection", new SimpleType(Types.COLLECTION, Types.STRING)),
        ("javaSet", new SimpleType(Types.SET, Types.STRING)),
        ("javaList", new SimpleType(Types.LIST, Types.STRING)),
        ("javaMap", new SimpleType(Types.MAP, Types.STRING, Types.STRING)),
        ("listOfPersons", new SimpleType(Types.LIST, entityType)),
        ("array", new ClassType(TypeCategory.ARRAY, classOf[Array[String]])),
        ("other", entityType))) {
      entityType.addProperty(new Property(entityType, name, t))
    }
        

  }

  @Test
  def Print {
    val typeMappings = ScalaTypeMappings.create
    typeMappings.register(entityType, new QueryTypeFactoryImpl("Q", "", "").create(entityType))
    val serializer = new ScalaEntitySerializer(typeMappings)
    serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new ScalaWriter(writer))
    val str = writer.toString()
    //System.err.println(str)
    assertTrue(str.contains("class QPerson(cl: Class[_ <: Person], md: PathMetadata[_]) " +
    		"extends EntityPathImpl[Person](cl, md) {"))
    assertTrue(str.contains("def this(variable: String) = " +
    		"this(classOf[Person], forVariable(variable))"))
    assertTrue(str.contains("def this(parent: Path[_], variable: String) = " +
    		"this(classOf[Person], forProperty(parent, variable))"))
  }

  @Test
  def Compile {
    val typeMappings = ScalaTypeMappings.create
    typeMappings.register(entityType, new QueryTypeFactoryImpl("Q", "", "").create(entityType))
    val serializer = new ScalaEntitySerializer(typeMappings)
    serializer.serialize(entityType, SimpleSerializerConfig.DEFAULT, new ScalaWriter(writer))
    val str = writer.toString()
    //System.err.println(str);
    assertCompileSuccess(str)
  }
}